apisix/discovery/consul/init.lua

```diff
@@ -0,0 +1,418 @@
+    }
+    local data = core.json.encode(entity)
+    local succ, err =  util.write_file(dump_params.path, data)
+    if not succ then
+        log.error("write dump into file got error: ", err)
+    end
+end
+
+
+local function show_dump_file()
+    if not dump_params then
+        return 503, "dump params is nil"
+    end
+
+    local data, err = util.read_file(dump_params.path)
+    if not data then
+        return 503, err
+    end
+
+    return 200, data
+end
+
+
+function _M.connect(premature, consul_server, retry_delay)
+    if premature then
+        return
+    end
+
+    local consul_client = resty_consul:new({
+        host = consul_server.host,
+        port = consul_server.port,
+        connect_timeout = consul_server.connect_timeout,
+        read_timeout = consul_server.read_timeout,
+        default_args = consul_server.default_args,
+    })
+
+    log.info("consul_server: ", json_delay_encode(consul_server, true))
+    local watch_result, watch_err = consul_client:get(consul_server.consul_watch_sub_url)
+    local watch_error_info = (watch_err ~= nil and watch_err)
+            or ((watch_result ~= nil and watch_result.status ~= 200)
+            and watch_result.status)
+    if watch_error_info then
+        log.error("connect consul: ", consul_server.consul_server_url,
+                " by sub url: ", consul_server.consul_watch_sub_url,
+                ", got watch result: ", json_delay_encode(watch_result, true),
+                 ", with error: ", watch_error_info)
+
+        if not retry_delay then
+            retry_delay = 1
+        else
+            retry_delay = retry_delay * 4
+        end
+
+        log.warn("retry connecting consul after ", retry_delay, " seconds")
+        core_sleep(retry_delay)
+
+        goto ERR
+    end
+
+    log.info("connect consul: ", consul_server.consul_server_url,
+            ", watch_result status: ", watch_result.status,
+            ", watch_result.headers.index: ", watch_result.headers['X-Consul-Index'],
+            ", consul_server.index: ", consul_server.index,
+            ", consul_server: ", json_delay_encode(consul_server, true))
+
+    -- if current index different last index then update service
+    if consul_server.index ~= watch_result.headers['X-Consul-Index'] then
+
+        -- fetch all services info
+        local result, err = consul_client:get(consul_server.consul_sub_url)
+
+        local error_info = (err ~= nil and err) or
+         ((result ~= nil and result.status ~= 200) and result.status)
+
+        if error_info then
+            log.error("connect consul: ", consul_server.consul_server_url,
+                    " by sub url: ", consul_server.consul_sub_url,
+                    ", got result: ", json_delay_encode(result, true),
+                    ", with error: ", error_info)
+
+            if not retry_delay then
+                retry_delay = 1
+            else
+                retry_delay = retry_delay * 4
```

Comment:
Can we wrap the get+retry into a function?

Reply:
done